Run st2-self-check in Github Actions

winem opened the pull request:
This PR enables us to run the st2-self-check as part of our CI run in Github Actions as well. 

The most crucial part was to handle the environment correctly and to determine the context when running the self-check. A few modifications are needed to adapt to the GHA environment. 

Another big change is that the st2-self-check exits with a return code != 0 if any of the tests fail. This is required to ensure that the CI run fails. 
I do not see a use-case where this could break anything on the users end but I'll add a note to the changelog to highlight this change as well. I don't think it's a breaking change - but let me know if you disagree. 

~All tests but 4 (2 remote actions that try to connect to localhost via ssh and 2 HTTP actions) succeed.~

~I'm open for any input on how to handle the remaining 4 tests.~

~@armab suggested to have a look at  https://github.com/StackStorm/st2packaging-dockerfiles/tree/master/packagingtest and see if we can re-use this here. I'll evaluate it and am open for further ideas.~
~It's the same for the HTTP actions.~

~I'll mark this PR as draft until we have a solution for the 4 failing tests but feel free to review the changes I already made and give your feedback!~

There may always be tests that may not succeed in GitHub Actions due to special infrastructure, environment or further external dependencies to be satisfied to make them work. So I added some kind of "skip list" or "black list" of tests. These tests will not be executed by GitHub Actions. This list is a space separated list of string configured as environment variable for the `self-check` job. So you don't have to handle this kind of exceptions in the test files itself and just add them in a single place.

As agreed in a previous TSC meeting the latest Python3.8 release is used. 

closes #5496

armab commented:
The idea with that container is that it comes with an init system and ssh and so you basically use it the same as a VM for testing.
We use it in CI Ansible-st2, Puppet-st2, and st2-packages integration testing.

winem commented:
I found a solution for the ssh issue and these actions work now. Now I'll see if I find any hint why the two HTTP actions fail.

winem commented:
The execution of the self-checks works as expected now. I'll add the CHANGELOG-text soon and will also have a look at the failing integration test.

winem commented:
As just agreed in the TSC I'll create issues to run the self-check on other Py versions as well. For now it's Python3.6 that's interesting but Python 3.9 or even 3.10 will become interesting "soon". 

There'll be another issue to run the self-check against different OS.

amanda11 commented on .github/workflows/ci.yaml:
```diff
@@ -136,6 +136,163 @@ jobs:
+        run: |
+          sudo ST2_CI_REPO_PATH="${ST2_CI_REPO_PATH}" scripts/ci/permissions-workaround.sh
+      - name: Reconfigure RabbitMQ
+        # bitnami image allows (see bitnami/rabbitmq readme):
```
I didn't quite understand the comment on the reconfigure rabbitMQ, could we rephrase - wasn't sure what the point was being made about the bitnami image allows, or the bit about needing to switch to custom.conf or advanced.config

winem replied:
I reviewed the file and to be honest I think they were probably left overs from the development phase. I removed the comment as well as 2 other ones (the 2 other ones due to infelicitous wording).

nzlosh commented on st2common/bin/st2-self-check:
```diff
@@ -136,6 +142,11 @@ do
         continue
     fi
 
+    if [[ -n "${GITHUB_ACTIONS}" && " ${TESTS_TO_SKIP_IN_GHA} " =~ " ${TEST} " ]]; then
```
+1

winem replied:
Also done and added a note in the changelog.

nzlosh reviewed (approved):
I like the idea of a generic skip variable, overall this looks good to me.

winem commented:
So I switched to a generic TESTS_TO_SKIP variable that can be used by anyone and added a note to the changelog. So, from my side, this PR is ready to be reviewed again.

armab reviewed (changes requested):
It works 🎉  Thanks for all the efforts!

A couple of minor comments before merging

winem commented on st2common/bin/st2-self-check:
```diff
@@ -136,6 +142,11 @@ do
         continue
     fi
 
+    if [[ -n "${GITHUB_ACTIONS}" && " ${TESTS_TO_SKIP} " =~ " ${TEST} " ]]; then
```
I just realized another glitch here. `-n "${GITHUB_ACTIONS}"` is no longer needed when we want to have a generic `TESTS_TO_SKIP`. I'll fix this.

winem replied:
Done.

winem commented on CHANGELOG.rst:
```diff
@@ -15,6 +15,9 @@ Fixed
 
 * Avoid logging sensitive information in debug (fix #5977)
 
+* Run the st2 self-check in Github Actions and support the environment variable `TESTS_TO_SKIP` to skip tests when running st2-self-check . #5609
```
I decided to just move the record to `Added` as per your suggestion and I removed the 2nd entry as it affects the behavior st2-self-check just in the context of Github Actions. And we didn't have the check executed via GHA before, so it's not really a change.

winem commented on .github/workflows/ci.yaml:
```diff
@@ -144,6 +144,163 @@ jobs:
+    needs: pre_job
+    # NOTE: We always want to run job on master since we run some additional checks there (code
+    # coverage, etc)
+    if: ${{ needs.pre_job.outputs.should_skip != 'true' || github.ref == 'refs/heads/master' }}
```
It absolutely makes sense to fail as early as possible. I removed the condition. Good find!

armab reviewed (approved):
Great work @winem! 
Thanks!
